File: Algorithms/BreadthFirstSearch/Graph.cpp
#include <iostream>
#include <map>
#include <vector>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <queue>

struct Edge
{
  int fr;
  int to;
  int wt;

  Edge(int fr, int to, int wt = 1)
    : fr(fr), to(to), wt(wt) { }

  Edge Copy()
  {
    return Edge(fr, to, wt);
  }
};

struct Graph
{
  std::unordered_map<int, std::vector<Edge>> adjacencyList;

  Graph() { }

  Graph(std::vector<Edge> edgelist)
  {
    for (Edge& edge : edgelist)
    {
      AddEdge(edge);
    }
  }

  void AddNode(int node)
  {
    if (adjacencyList.find(node) != adjacencyList.end()) return;
    adjacencyList[node] = std::vector<Edge>();
  }

  void AddEdge(Edge& edge)
  {
    AddNode(edge.fr);
    AddNode(edge.to);
    adjacencyList[edge.fr].emplace_back(edge.Copy());
  }

  void PrintAdjacencyList()
  {
    for (std::pair<int, std::vector<Edge>> pair : adjacencyList)
    {
      std::cout << pair.first << " : ";
      for (Edge edge : pair.second)
      {
        std::cout << edge.to << "(" << edge.wt << ")" << " -> ";
      }
      std::cout << "X\n";
    }
  }

  std::vector<int> GetBFSSequence()
  {
    int first;
    for (auto pair : adjacencyList) 
    {
      first = pair.first;
      break;
    }
    return GetBFSSequence(first);
  }

  std::vector<int> GetBFSSequence(int root)
  {
    std::vector<int> sequence;

    if (adjacencyList.begin() == adjacencyList.end()) {
      return sequence;
    }

    std::queue<int> q;
    for (std::pair<int, std::vector<Edge>> pair : adjacencyList) {
      q.push(pair.first);
      break;
    }

    while (!q.empty()) {
      if (std::find(sequence.cbegin(), sequence.cend(), q.front()) != sequence.cend()) {
        q.pop();
        continue;
      }
      sequence.emplace_back(q.front());
      for (Edge edge : adjacencyList[q.front()]) {
        if (std::find(sequence.cbegin(), sequence.cend(), edge.to) == sequence.cend())
          q.push(edge.to);
      }
      q.pop();
    }

    return sequence;
  }
};

int main()
{
  std::vector<Edge> edgelist = {
    Edge(1, 2),
    Edge(1, 3),
    Edge(1, 4),
    Edge(2, 3),
    Edge(3, 4),
    Edge(3, 5),
    Edge(4, 5),
  };

  Graph gr(edgelist);

  gr.PrintAdjacencyList();

  for (int val : gr.GetBFSSequence())
  {
    std::cout << val << " ";
  }
  std::cout << "\n";

  return 0;
}

File: Algorithms/BreadthFirstSearch/Graph.py
from typing import *;
from typing_extensions import *;
import queue;

class Edge:
  def __init__(self:Self, fr:int, to:int, wt:int = 1) -> None:
    self.fr:int = fr;
    self.to:int = to;
    self.wt:int = wt;

class Graph:
  def __init__(self:Self, edgelist:List[Edge]|None = None) -> None:
    self.adjacencyList:Dict[int,Set[Edge]] = dict()
    if edgelist == None: return;
    for edge in edgelist:
      self.AddNode(edge.fr);
      self.AddNode(edge.to);
      self.AddEdge(edge);
    return;

  def AddNode(self:Self, node:int) -> None:
    if node in self.adjacencyList: return;
    self.adjacencyList[node] = set();
    return;

  def AddEdge(self:Self, edge:Edge) -> None:
    self.adjacencyList[edge.fr].add(edge);
    return;

  def PrintAdjacencyList(self:Self) -> None:
    for element in self.adjacencyList:
      print(element, sep="", end=" -> ");
      for edge in self.adjacencyList[element]: print(f"{edge.to}({edge.wt})", sep="", end="")
      print("X");
    return;

  def BFS(self:Self, root:int|None = None) -> List[int]:
    sequence:List[int] = [];

    if not self.adjacencyList:
      return sequence;

    q:queue.Queue[int] = queue.Queue();
    q.put(root or list(self.adjacencyList.keys())[0]);
    
    while not q.empty():
      cur:int = q.get();
      if cur in sequence:
        continue;
      sequence.append(cur);
      for edge in self.adjacencyList[cur]:
        if edge.to not in sequence:
          q.put(edge.to);

    return sequence;


def main() -> None:
  edgelist = [
    Edge(1, 2),
    Edge(1, 3), 
    Edge(1, 4),
    Edge(2, 3),
    Edge(3, 4),
    Edge(3, 5),
    Edge(4, 5),
  ];

  gr = Graph(edgelist);

  gr.PrintAdjacencyList();

  print(gr.BFS());

if __name__ == "__main__": main();